File: general-interview-prep/golang/minDepthBinaryTree.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


 // [3,9,20,null,null,15,7],
 //
 //    3
 //   / \
 //  9  20
 //    /  \
 //   15   7


type NodeDequeue struct {
    Depth int
    Node *TreeNode
}

func minDepth(root *TreeNode) int {
    // Null node has 0 depth.
    if root == nil {
        return 0
    }

    // Initialize a list to be used a queue with the root.
    q := make([]*NodeDequeue, 0)
    q = append(q, &NodeDequeue{Node: root, Depth: 1})

    // Level order traversal algorithm:
    for len(q) > 0 {
        // Get the node at the front of the queue.
        queueItem := q[0]
        q = q[1:]

        // Retreive the data that the node had
        node := queueItem.Node
        depth := queueItem.Depth

        // If this is the first leaf node encountered,
        // return its depth and terminate the algorithm.
        if node.Left == nil && node.Right == nil {
          return depth
        }

        // If left subtree exists, add it to queue
        if node.Left != nil {
            q = append(q, &NodeDequeue{Node: node.Left , Depth: depth + 1})
        }

        // if right subtree exists, add it to queue
        if node.Right != nil {
            q = append(q, &NodeDequeue{Node: node.Right, Depth: depth + 1})
        }
    }
    return 0
}

func main() {

}

File: general-interview-prep/javascript/minDepthBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
 var minDepth = function(root) {
  if (!root) {
      return 0;
  }
  
  let queue = [root];
  let node = root;
  let depth = 0;
  while (queue.length) {
      let queueLen = queue.length;
      depth++;
      for(var i = 0; i < queueLen; i++) {
          node = queue.shift();
          if(!node.left && !node.right) return depth;
          if(node.left) queue.push(node.left);
          if(node.right) queue.push(node.right);
      }
  }
  
  return depth;
};